### PR TITLE
Handling "no profile found" case

### PR DESCRIPTION
**Pull request by rkzel**
One weird thing about this one is that 3box client does not exactly inform about profile being non-existent (despite the fact that https://3box/$profile fails with simple 404 internally) - it just returns {}. I used that to set noPublicProfileFound flag in our state to make appropriate screen show up.

### CODE REVIEW

**Review by Schwartz10 — CHANGES_REQUESTED** [review]
@rkzel this is an interesting approach to detecting no profiles! Instead, I think there's a much simpler approach that's already built in - by using the [`hasProfile`](https://github.com/AutarkLabs/aragon-profile/blob/master/src/modules/3box-aragon.js#L134) method. I get concerned about changing the `getPublic` method - it's been very carefully crafted lol :)

I really like how you updated `useProfile` hook to have a profile method - so let's keep that _idea_ and update the logic to call `hasProfile` instead of creating a new state variable

**Comment by rkzel** [comment]
I've looked into hasProfile tried that approach. Two things:

1. hasProfile makes its own call to 3box to determine whether profile exists, combined with getPublic call that's going to happen anyway we get this (please ignore my test console messages):

Screenshot from 2019-07-09 13-59-34

So what I did at the deepest level was determining whether the profile exists while it is being fetched and set state accordingly. The reason for using state is basically this:

`if (Object.keys(error).length > 0) return <ErrorState />`
`if (isInitializing) return <Initializing />`
`if (isLoadingPublicProfile) return <LoadingPublicProfile />`
`if (isUnlockingProfile) return <UnlockingBox />`
`if (noPublicProfileFound) return <NoProfile ethereumAddress={ethereumAddress} />`
`return <Fragment>{children}</Fragment>`

That's in LoadAndErrorWrapper. State variables determine everything, so it felt more natural than for example:

`if (box.hasProfile()) return <NoProfile />`
`if (isInitializing) return <Initializing />`
`if (isLoadingPublicProfile) return <LoadingPublicProfile />`
...

It would be possible to skip <LoadAndErrorWrapper> entirely in case there is no profile to load and show, but that too would be slightly artificial :)

**Comment by Schwartz10** [comment]
@rkzel good point about not making unnecessary http requests if we don't need to - let's stay away from the `hasProfile` method. A couple things:

I think the `LoadAndErrorWrapper` is a good place to put the logic for taking the information about whether or not the user has a profile and rendering the `NoProfile` view. 

I still don't think we need to add another state variable + action creators + reducer handlers for `hasProfile` - the reason being is that since `LoadAndErrorWrapper` is the single place in the UI that's going to consume the `hasProfile` information, we can just calculate this in the `useProfileStates` hook. This will be a good opportunity to clean up the LoadAndErrorWrapper to use the `useProfileStates` hook too instead of re-implementing a lot of logic.

So in the `LoadAndErrorView` we can consume state like:

```js
const { loadingPublicProf, unlockingProf, hasProfile } = useProfileStates()

  return (
    <LoadAndErrorView
      {...props}
      isInitializing={syncing}
      hasProfile={hasProfile}
      isLoadingPublicProfile={loadingPublicProf}
      isUnlockingProfile={unlockingProf}
    />
  )
```

Does this make sense?

**Comment by Schwartz10** [comment]
Also, with the design, how does one actually go about creating a profile from the no profile page? Does a no profile page look different for a user viewing their own no profile page versus viewing someone else's?

File: src/modules/3box-aragon.js [hunk 1]
```diff
@@ -62,12 +62,14 @@ export class Profile {
     const publicProfile = this.boxState.opened
       ? await this.unlockedBox.public.all()
       : await Box.getProfile(this.ethereumAddress)
-
     return publicProfile
   }
 
   getPublic = async () => {
     const publicProfile = await this._getPublic()
+    if (!Object.keys(publicProfile).length) {
```
Comment (Schwartz10) [inline src/modules/3box-aragon.js:70]: Why is this necessary?

File: src/client/wrappers/loadAndErrorWrapper/index.js [hunk 2]
```diff
@@ -1,78 +1,37 @@
...
-  isUnlockingProfile,
-  error,
-}) => {
   if (Object.keys(error).length > 0) return <ErrorState />
   if (isInitializing) return <Initializing />
-  if (isLoadingPublicProfile) return <LoadingPublicProfile />
-  if (isUnlockingProfile) return <UnlockingBox />
+  if (loadingPublicProf) return <LoadingPublicProfile />
+  if (unlockingProf) return <UnlockingBox />
+  if (noPublicProfileFound) {
```
Comment (Schwartz10) [inline src/client/wrappers/loadAndErrorWrapper/index.js]: please remove brackets here for consistency :)

File: src/client/wrappers/loadAndErrorWrapper/index.js [hunk 3]
```diff
@@ -1,78 +1,37 @@
...
+  const {
+    loadingPublicProf,
+    unlockingProf,
+    noPublicProfileFound,
+    error,
+  } = useProfileStates()
 
-  const { boxes } = useContext(BoxContext)
+  // it was hardcoded as syncing=false before
+  const isInitializing = false
```
Comment (Schwartz10) [inline src/client/wrappers/loadAndErrorWrapper/index.js:20]: Let's fix this in this PR - this should be taken from the `syncing` variable given from the`useAragonApi` hook

File: src/client/stateManagers/box/states.js [hunk 4]
```diff
@@ -78,6 +79,15 @@ export const fetchedPublicProfileErr = (state, error) => ({
   error,
 })
 
+export const noPublicProfileFound = (state, error) => ({
+  ...state,
+  loadingPublicProf: false,
+  loadedPublicProf: false,
+  loadedPublicProfSuccess: false,
```
Comment (Schwartz10) [inline src/client/stateManagers/box/states.js]: Think we should set these state variables as though the profile was loaded 

loading: false
loaded: true
loadedSuccess: true

**Comment by rkzel** [comment]
> Also, with the design, how does one actually go about creating a profile from the no profile page? Does a no profile page look different for a user viewing their own no profile page versus viewing someone else's?

Design says all the fields should be visible if the not-yet-existing profile belongs to the user viewing it. I added it now.

File: src/client/wrappers/loadAndErrorWrapper/index.js (hunk 2 above)
Comment (rkzel) [reply src/client/wrappers/loadAndErrorWrapper/index.js]: Sure!

File: src/client/stateManagers/box/states.js (hunk 4 above)
Comment (rkzel) [reply src/client/stateManagers/box/states.js]: Makes sense.

File: src/client/wrappers/loadAndErrorWrapper/index.js (hunk 3 above)
Comment (rkzel) [reply src/client/wrappers/loadAndErrorWrapper/index.js:20]: I tried, including installing api packages, but useAragonApi expects to live in an application that communicates with Wrapper, not one that is inserted straight into it?

File: src/modules/3box-aragon.js (hunk 1 above)
Comment (rkzel) [reply src/modules/3box-aragon.js:70]: use3Box hook gets Profile like this:

`const publicProfile = await profile.getPublic()`

getPublic() downloads profile and adds github and twitter keys to it. If Profile is empty - return early, so the fields are not added, and the hook knows there is no profile. There are probably two alternatives here: either check if getPublic() returns empty object, or if returns an object missing one special key that should be there and if there isn't it means the profile is empty.

File: src/client/stateManagers/box/states.js (hunk 4 above)
Comment (Schwartz10) [reply src/client/stateManagers/box/states.js]: `loadedPublicProfileSuccess` i think should be marked true - because this state appears to be an error

Comment (Schwartz10) [reply src/client/stateManagers/box/states.js]: @rkzel this is still not fixed

**Review by Schwartz10 — CHANGES_REQUESTED** [review]
See the one small state change, then this should be good to go. I'll test it one more time live before we merge after the fix is made.

**Review by Schwartz10 — APPROVED** [review]

File: src/client/wrappers/loadAndErrorWrapper/index.js [hunk 5]
```diff
@@ -1,78 +1,37 @@
...
+const LoadAndErrorWrapper = ({ children, ethereumAddress }) => {
+  const {
+    loadingPublicProf,
+    unlockingProf,
+    noPublicProfileFound,
+    error,
+  } = useProfileStates()
+  const { viewMode } = useProfile()
+
+  // it was hardcoded as syncing=false before
```
Comment (chadoh) [inline src/client/wrappers/loadAndErrorWrapper/index.js]: This comment seems like it's relevant to this PR, but like it won't be useful in the codebase beyond that point. I'd rather it didn't get committed. Instead, I'd prefer a PR line comment (such as this one).

File: src/client/wrappers/loadAndErrorWrapper/NoProfile.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+      <div css="display: flex; margin-bottom: 24px">
+        <IconEthereum width="18px" height="18px" color={theme.textTertiary} />
+        <Text
+          css="font-size: 13px; margin-left: 8px"
+          color={theme.textTertiary}
+        >
+          {ethereumAddress}
+        </Text>
+      </div>
+      <Text.Block css="font-size: 15px">
```
Comment (chadoh) [inline src/client/wrappers/loadAndErrorWrapper/NoProfile.js]: Here and above, can we use the closest matching `size` prop for `Text`, rather than using custom sizes? I'd rather match the official Aragon design system than our specific designs. That will result in greater consistency with the rest of Aragon, and when aragonUI gets updated, we can be right in sync with it.

**Review by chadoh — APPROVED** [review]
A couple small notes. No need for me to re-review after these are addressed.

File: src/client/wrappers/loadAndErrorWrapper/index.js (hunk 5 above)
Comment (Schwartz10) [reply src/client/wrappers/loadAndErrorWrapper/index.js]: @rkzel can you make a github ticket out of this and put in in the backlog?

Comment (rkzel) [reply src/client/wrappers/loadAndErrorWrapper/index.js]: Fixed.

File: src/client/wrappers/loadAndErrorWrapper/NoProfile.js (hunk 6 above)
Comment (rkzel) [reply src/client/wrappers/loadAndErrorWrapper/NoProfile.js]: 15px happens to be "normal", which is also the default... fixed.

File: src/client/stateManagers/box/states.js (hunk 4 above)
Comment (rkzel) [reply src/client/stateManagers/box/states.js]: `useLinkedData` will try to format public profile if `loadedPublicProfileSuccess` is set to true.
